src/util/test_utils.rs

```diff
@@ -0,0 +1,621 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+
+use crate::blockdata::transaction::{ExtraField, KeyImage, RawExtraField, SubField, TxOutTarget};
+use crate::consensus::encode::Encodable;
+use crate::consensus::{deserialize, serialize};
+use crate::cryptonote::hash::Hashable;
+use crate::util::key::H;
+use crate::util::ringct::{CtKey, EcdhInfo, Key, RctSig, RctSigBase, RctSigPrunable, RctType};
+use crate::{
+    Amount, Block, BlockHeader, Hash, PrivateKey, PublicKey, Transaction, TransactionPrefix, TxIn,
+    TxOut, VarInt, ViewPair,
+};
+use hex::{FromHex, ToHex};
+use std::io;
+use std::str::FromStr;
+
+/// Fuzz for block deserialization, called from the fuzz target
+pub fn fuzz_block_deserialize(fuzz_data: &[u8]) -> bool {
+    let fuzz_bytes = fuzz_data.to_vec();
+
+    // Block
+    if let Ok(val) = deserialize::<Block>(&fuzz_bytes[..]) {
+        assert_eq!(fuzz_bytes, serialize(&val));
+    }
+
+    true
+}
+
+/// Fuzz for block header deserialization, called from the fuzz target
+pub fn fuzz_block_header_deserialize(fuzz_data: &[u8]) -> bool {
+    let fuzz_bytes = fuzz_data.to_vec();
+
+    // BlockHeader
+    if let Ok(val) = deserialize::<BlockHeader>(&fuzz_bytes[..]) {
+        if fuzz_bytes != serialize(&val) {
+            println!("fuzz_data: {:?}", fuzz_data);
+        }
+        assert_eq!(fuzz_bytes, serialize(&val));
+    }
+
+    true
+}
+
+/// Fuzz for transaction prefix deserialization, called from the fuzz target
+pub fn fuzz_transaction_prefix_deserialize(fuzz_data: &[u8]) -> bool {
+    let fuzz_bytes = fuzz_data.to_vec();
+
+    // TransactionPrefix
+    if let Ok(val) = deserialize::<TransactionPrefix>(&fuzz_bytes[..]) {
+        assert_eq!(fuzz_bytes, serialize(&val));
+    }
+
+    true
+}
+
+/// Padding sub-field position in extra data
+#[derive(Clone, Copy)]
+pub enum AddPadding {
+    /// Add padding to the front of the extra field
+    ToFront,
+    /// Add padding to the middle of the extra field
+    ToMiddle,
+    /// Add padding to the rear of the extra field
+    ToRear,
+}
+
+/// Fuzz helper function to create an extra field, called from the fuzz target
+pub fn fuzz_create_extra_field(fuzz_data: &[u8], add_padding: AddPadding) -> ExtraField {
+    let fuzz_bytes = fuzz_data.to_vec();
+    let hash = Hash::new(fuzz_data);
+
+    // SubField::TxPublicKey
+    let tx_pub_key_field = SubField::TxPublicKey(match PublicKey::from_slice(fuzz_data) {
+        Ok(val) => val,
+        Err(_) => H,
+    });
+    let nonce_field = SubField::Nonce(fuzz_bytes.clone());
+
+    // SubField::Padding
+    let padding_field = if fuzz_bytes.is_empty() {
+        SubField::Padding(u8::MIN)
+    } else {
+        SubField::Padding(fuzz_bytes[0])
```

Comment:
The purpose of this function is to create a randomized `ExtraField`, so the downstream code can be evaluated for panics or errors. We do not want it to be always valid.

Reply:
IMO if someone is making extra fields and they are doing it incorrectly it is their fault if they get an error/ panic.

Plus if we don't enforce the padding we have to do checks on the error if we had a middle or start padding and then we ignore the error but the error could have been for something unrelated to the padding

Reply:
The way the test has been written (`pub fn fuzz_extra_field_try_parse(`) compensates for the way the decoding has been implemented. With the ser -> des roundtrip we only panic on a parsing error when we definitely know decoding should have been successful, and ignore a parsing error otherwise. We need to exercise all modes and the test input data is allowed to do that.